R-package/R/xgboost.R

```diff
@@ -1,32 +1,1022 @@
+            objective
+          )
+        }
+      }
+
+      if (!is.null(base_margin)) {
+        base_margin <- process.base.margin(base_margin, length(y), length(y_levels))
+      }
+
+      out <- list(
+        params = list(
+          objective = objective,
+          num_class = length(y_levels)
+        ),
+        metadata = list(
+          y_levels = y_levels,
+          n_targets = length(y_levels)
+        )
+      )
+    }
+
+    out$dmatrix_args <- list(
+      label = as.numeric(y) - 1,
+      base_margin = base_margin
+    )
+
+  } else if (inherits(y, "Surv")) {
+
+    y_attr <- attributes(y)
+    supported_surv_types <- c("left", "right", "interval")
+    if (!(y_attr$type %in% supported_surv_types)) {
+      stop(
+        "Survival objectives are only supported for types: ",
+        paste(supported_surv_types, collapse = ", "),
+        ". Was passed: ",
+        y_attr$type
+      )
+    }
+
+    if (is.null(objective)) {
+      objective <- "survival:aft"
+    } else {
+      if (y_attr$type == "right") {
+        if (!(objective %in% .SURVIVAL_RIGHT_CENSORING_OBJECTIVES())) {
+          stop(
+            "Got right-censored 'y' variable - supported objectives for this data are: ",
+            paste(.SURVIVAL_RIGHT_CENSORING_OBJECTIVES(), collapse = ", "),
+            ". Was passed: ",
+            objective
+          )
+        }
+      } else {
+        if (!(objective %in% .SURVIVAL_ALL_CENSORING_OBJECTIVES())) {
+          stop(
+            "Got ", y_attr$type, "-censored 'y' variable - supported objectives for this data are:",
+            paste(.SURVIVAL_ALL_CENSORING_OBJECTIVES(), collapse = ", "),
+            ". Was passed: ",
+            objective
+          )
+        }
+      }
+    }
+
+    if (!is.null(base_margin)) {
+      base_margin <- process.base.margin(base_margin, nrow(y), 1)
+    }
+
+    out <- list(
+      params = list(
+        objective = objective
+      ),
+      metadata = list(
+        n_targets = 1
+      )
+    )
+
+    if (objective == "survival:cox") {
+      # Can only get here when using right censoring
+      if (y_attr$type != "right") {
+        stop("Internal error.")
+      }
+
+      out$dmatrix_args <- list(
+        label = y[, 1L] * (2 * (y[, 2L] - 0.5))
+      )
+
+    } else {
+      if (y_attr$type == "left") {
+        lb <- ifelse(
+          y[, 2L] == 0,
+          0,
+          y[, 1L]
+        )
+        ub <- y[, 1L]
+        out$dmatrix_args <- list(
+          label_lower_bound = lb,
+          label_upper_bound = ub
+        )
+      } else if (y_attr$type == "right") {
+        lb <- y[, 1L]
+        ub <- ifelse(
+          y[, 2L] == 0,
+          Inf,
+          y[, 1L]
+        )
+        out$dmatrix_args <- list(
+          label_lower_bound = lb,
+          label_upper_bound = ub
+        )
+      } else if (y_attr$type == "interval") {
+        out$dmatrix_args <- list(
+          label_lower_bound = ifelse(y[, 3L] == 2, 0, y[, 1L]),
```

Comment:
Could you please add some comments for these indexing?

Reply:
Added a comment block above.